### PR TITLE
zydis: split zycore into separate derivation

### PR DESCRIPTION
**Pull request by athre0z**
###### Description of changes

This PR splits the Zycore dependency that was previously pulled in via `fetchSubmodules` into a separate derivation. The package previously didn't install the Zycore headers and CMake configs, making it impossible to actually compile any code that uses Zydis because Zydis includes Zycore headers. 

Essentially, the `tl;dr` here is:

- CMake's `install` rule doesn't install headers and config for dependencies
- Zycore is a separate package with separate versions
- Zydis headers `#include`s Zycore headers which are currently not installed

Instead of adding code that manually install the missing files somehow, I figured that it'd probably be most idiomatic to properly build it as a separate derivation. Zycore isn't widely used for anything but a helper to Zydis, so I didn't put it into a completely separate package. I don't feel strongly about that and am also perfectly open to making it a top-level package instead, though I don't think that many people will ever use it.

I also added myself as a maintainer for the package.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux (via qemu)
  - [x] x86_64-darwin (via Rosetta)
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

CC @jbcrail @AndersonTorres

### CODE REVIEW

File: pkgs/development/libraries/zydis/default.nix [hunk 1]
```diff
@@ -2,8 +2,31 @@
 , stdenv
 , fetchFromGitHub
 , cmake
+, python3
 }:
 
+let
+  zycore = stdenv.mkDerivation rec {
```
Comment (AndersonTorres) [inline pkgs/development/libraries/zydis/default.nix:9]: It can be put in its own file.

File: pkgs/development/libraries/zydis/default.nix [hunk 2]
```diff
@@ -12,19 +35,33 @@ stdenv.mkDerivation rec {
...
+    "-DZYAN_SYSTEM_ZYCORE=ON"
+    "-DCMAKE_INSTALL_LIBDIR=lib"
+    "-DCMAKE_INSTALL_INCLUDEDIR=include"
   ];
 
+
+  doCheck = true;
+  checkInputs = [ python3 ];
+  checkPhase = ''
+    (cd ../tests && python3 ./regression.py test ../build/ZydisInfo)
```
Comment (AndersonTorres) [inline pkgs/development/libraries/zydis/default.nix:53]: Use pushd / popd instead of subshell.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by athre0z** [comment]
Thanks for the swift review! I addressed both comments.

**Review by AndersonTorres — APPROVED** [review]

**Comment by ehamberg** [comment]
Hey, @athre0z. Sorry for digging up something this old, but I'm trying to build a project with a `find_package(zydis CONFIG REQUIRED)` in their `CMakeLists.txt` and after adding `zydis` to the nix dev environment, I get this error when running `cmake`:

```
CMake Error at /nix/store/rv1sx6lj8x5ffik7vzy0m3w0a4ci6pnx-cmake-3.29.6/share/cmake-3.29/Modules/CMakeFindDependencyMacro.cmake:76 (find_package):
  By not providing "FindZycore.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "Zycore", but
  CMake did not find one.

  Could not find a package configuration file provided by "Zycore" with any
  of the following names:

    ZycoreConfig.cmake
    zycore-config.cmake

  Add the installation prefix of "Zycore" to CMAKE_PREFIX_PATH or set
  "Zycore_DIR" to a directory containing one of the above files.  If "Zycore"
  provides a separate development package or SDK, be sure it has been
  installed.
Call Stack (most recent call first):
  /nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/lib/cmake/zydis/zydis-config.cmake:30 (find_dependency)
  CMakeLists.txt:8 (find_package)
```

Should there have been a `ZycoreConfig.cmake`/`zycore-config.cmake` file installed as part of the `zydis` package?

<details>
  <summary><code>$ find  /nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/</code></summary>
<code>/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/bin
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/bin/ZydisInfo
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/bin/ZydisDisasm
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Generated
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Generated/EnumRegister.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Generated/EnumInstructionCategory.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Generated/EnumISAExt.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Generated/EnumMnemonic.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Generated/EnumISASet.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Utils.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Status.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Disassembler.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/SharedTypes.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Segment.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Mnemonic.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Zydis.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Formatter.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Decoder.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Internal
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Internal/DecoderData.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Internal/FormatterIntel.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Internal/FormatterBase.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Internal/EncoderData.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Internal/SharedData.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Internal/FormatterATT.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Internal/String.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Defines.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/DecoderTypes.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Encoder.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/ShortString.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/Register.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/MetaInfo.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/include/Zydis/FormatterBuffer.h
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/lib
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/lib/libZydis.a
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/lib/cmake
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/lib/cmake/zydis
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/lib/cmake/zydis/zydis-config-version.cmake
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/lib/cmake/zydis/zydis-config.cmake
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/lib/cmake/zydis/zydis-targets-release.cmake
/nix/store/n1x4n3bw11j8vkjgs2i5byi4vh4k6waz-zydis-4.1.0/lib/cmake/zydis/zydis-targets.cmake
</code>
</details>

**Comment by ehamberg** [comment]
> can you try whether adding the Zydis dependency with `(pkgs.zydis.overrideAttrs (_: _: { propagatedBuildInputs = [pkgs.zydis.zycore]; })) r`esolves the issue?

It did! Thanks!

**Pull request merged**